File: backend/app/schemas/budget.py
from datetime import datetime
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, Field


class BudgetBase(BaseModel):
    amount: Decimal = Field(..., ge=0, description="预算金额")


class BudgetCreate(BudgetBase):
    store_id: int = Field(..., gt=0, description="门店ID")
    expense_type_id: int = Field(..., gt=0, description="费用科目ID")
    year: int = Field(..., ge=2000, le=2100, description="年份")
    month: int = Field(..., ge=1, le=12, description="月份")


class BudgetItemCreate(BaseModel):
    expense_type_id: int = Field(..., gt=0, description="费用科目ID")
    amount: float = Field(..., ge=0, description="预算金额")


class BudgetBatchCreate(BaseModel):
    store_id: int = Field(..., gt=0, description="门店ID")
    year: int = Field(..., ge=2000, le=2100, description="年份")
    month: int = Field(..., ge=1, le=12, description="月份")
    items: list[BudgetItemCreate] = Field(..., min_length=1, description="预算条目")


class BudgetUpdate(BudgetBase):
    pass


class BudgetSchema(BudgetBase):
    id: int
    store_id: int
    expense_type_id: int
    year: int
    month: int
    created_at: datetime
    updated_at: datetime
    created_by_id: Optional[int]
    updated_by_id: Optional[int]

    class Config:
        from_attributes = True


class BudgetAnalysisItem(BaseModel):
    expense_type_id: int
    expense_type_name: str
    budget_amount: float
    actual_amount: float
    variance: float  # 差异额 = 实际 - 预算
    variance_rate: float  # 差异率 = (实际 - 预算) / 预算 * 100
    is_over_budget: bool  # 是否超支


class BudgetAnalysisResponse(BaseModel):
    total_budget: float
    total_actual: float
    total_variance: float
    items: list[BudgetAnalysisItem]

File: apps/web/src/types/modules/budget.ts
export interface Budget {
  id: number
  store_id: number
  expense_type_id: number
  year: number
  month: number
  amount: number
  created_at?: string
  updated_at?: string
}

export interface BudgetItemInput {
  expense_type_id: number
  amount: number
}

export interface BudgetBatchCreate {
  store_id: number
  year: number
  month: number
  items: BudgetItemInput[]
}

export interface BudgetAnalysisItem {
  expense_type_id: number
  expense_type_name: string
  budget_amount: number
  actual_amount: number
  variance: number        // 差异额 = 实际 - 预算
  variance_rate: number   // 差异率 %
  is_over_budget: boolean // 是否超支
}

export interface BudgetAnalysisResponse {
  total_budget: number
  total_actual: number
  total_variance: number
  items: BudgetAnalysisItem[]
}

export interface BudgetQuery {
  store_id: number
  year: number
  month: number
}
